setup.py

```diff
@@ -15,17 +15,50 @@
 __version__ = '4.4.1'
 
 import os
-import platform
-import sys
+
+from distutils.errors import CCompilerError
+from distutils.errors import DistutilsExecError
+from distutils.errors import DistutilsPlatformError
 
 from setuptools import Extension
 from setuptools import find_packages
 from setuptools import setup
+from setuptools.command.build_ext import build_ext
+
+
+def _read(fname):
+    here = os.path.abspath(os.path.dirname(__file__))
+    with open(os.path.join(here, fname)) as f:
+        return f.read()
+
+
+README = _read("README.rst") + '\n\n' + _read('CHANGES.rst')
+
 
-here = os.path.abspath(os.path.dirname(__file__))
-README = (open(os.path.join(here, 'README.rst')).read()
-          + '\n\n' +
-          open(os.path.join(here, 'CHANGES.rst')).read())
+class optional_build_ext(build_ext):
+    """This class subclasses build_ext and allows
+       the building of C extensions to fail.
+    """
+    def run(self):
+        try:
+            build_ext.run(self)
+        except DistutilsPlatformError as e:
+            self._unavailable(e)
+
+    def build_extension(self, ext):
+        try:
+            build_ext.build_extension(self, ext)
+        except (CCompilerError, DistutilsExecError, OSError) as e:
+            self._unavailable(e)
+
+    def _unavailable(self, e):
+        print('*' * 80)
+        print("""WARNING:
+        An optional code optimization (C extension) could not be compiled.
+        Optimizations for this package will not be available!""")
+        print()
+        print(e)
+        print('*' * 80)
```

Comment:
If you `print()` on Python 2, it'll print the repr of the empty tuple.

Please either use `print("")` or `from __future__ import print_function`.

Reply:
One thing I remember about `pip` is that it hides the output of `setup.py` , so end-users will not know their cached wheels lack C modules...

It would be nice if we could make `python setup.py bdist_wheel` **fail** unless PURE_PYTHON is set, in which case
1. pip will print the failure message to the user
2. pip will not cache a C-module-less wheel in ~/.cache
3. pip will fall back to running `python setup.py install` directly